File: src/main/java/net/charno/semweb/lookups/GeonamesClient.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package net.charno.semweb.lookups;

import net.charno.semweb.bindings.GeonamesBinding;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import net.charno.utils.StringUtils;
import org.apache.http.HttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.DefaultHttpClient;
import org.apache.http.util.EntityUtils;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

/**
 *
 * @author mdc502
 */
public class GeonamesClient extends SparqlClient{
    
    private static String URL_PRECISE = "http://api.geonames.org/search?name_equals=";
    private static String URL_FUZZY = "http://api.geonames.org/search?name=";
    private static String VAR_ROWS = "&maxRows=";
    private static String VAR_COUNTRY = "&country=";
    private static String VAR_FAVOR_UK = "&countryBias=GB";
    private static String VAR_CONTINENT = "&continentCode=";
    private static String VAR_SUFFIX = "&featureClass=A&featureClass=P&style=SHORT&type=json&username=";
    private static String URI = "http://sws.geonames.org/";
    
    private static Map<String,GeonamesBinding> parseJSON(String json) {
        Map<String,GeonamesBinding> results = new HashMap<String, GeonamesBinding>();
        JSONParser parser = new JSONParser();
        // process json results
        try {
            JSONObject jsonObject = (JSONObject) parser.parse(json);
            JSONArray jsonGeonames = (JSONArray) jsonObject.get("geonames");
            for (Object binding : jsonGeonames) {
                JSONObject jsonBinding = (JSONObject) binding;
                Long geonameId = (Long) jsonBinding.get("geonameId");
                String toponymName = (String) jsonBinding.get("toponymName");
                String countryCode = (String) jsonBinding.get("countryCode");
                String fcl = (String) jsonBinding.get("fcl");
                String fcode = (String) jsonBinding.get("fcode");
                
                GeonamesBinding geonamesBinding = new GeonamesBinding();
                geonamesBinding.setToponymName(toponymName);
                geonamesBinding.setCountryCode(countryCode);
                geonamesBinding.setFunctionClass(fcl);
                geonamesBinding.setFunctionCode(fcode);
                String uri = URI + geonameId;
                results.put(uri, geonamesBinding);
            }
        } catch (ParseException pe) {
            System.out.println("position: " + pe.getPosition());
            System.out.println(pe);
        }
        
        return results;
    }
    
    public static Map<String,GeonamesBinding> lookupPreciseLocationInCountry(String location, String country, int results, String username) {
        // parameter check
        if (results < 0 || results > 100) {
            results = 10;
        }
        // build SPARQL query
        String url = URL_PRECISE + StringUtils.makeURLSafe(location) + VAR_COUNTRY + country + VAR_ROWS + results + VAR_SUFFIX + username;
        // send query
        String json = getResultsFromEndpoint(url);
        
        return parseJSON(json);
    }
    
    public static Map<String,GeonamesBinding> lookupPreciseLocationInContinent(String location, String continent, int results, String username) {
        // parameter check
        if (results < 0 || results > 100) {
            results = 10;
        }
        // build SPARQL query
        String url = URL_PRECISE + StringUtils.makeURLSafe(location) + VAR_CONTINENT + continent + VAR_ROWS + results + VAR_SUFFIX + username;
        // send query
        String json = getResultsFromEndpoint(url);
        
        return parseJSON(json);
    }
    
    public static Map<String,GeonamesBinding> lookupPreciseLocationInWorld(String location, int results, String username) {
        // parameter check
        if (results < 0 || results > 100) {
            results = 10;
        }
        // build SPARQL query
        String url = URL_PRECISE + StringUtils.makeURLSafe(location) + VAR_ROWS + results + VAR_SUFFIX + username;
        // send query
        String json = getResultsFromEndpoint(url);
        
        return parseJSON(json);
    }
    
    public static Map<String,GeonamesBinding> lookupLocationFavorUK(String location, int results, String username) {
        // parameter check
        if (results < 0 || results > 100) {
            results = 10;
        }
        // build SPARQL query
        String url = URL_PRECISE + StringUtils.makeURLSafe(location) + VAR_FAVOR_UK + VAR_ROWS + results + VAR_SUFFIX + username;
        // send query
        String json = getResultsFromEndpoint(url);
        
        return parseJSON(json);
    }
    
    public static Map<String,GeonamesBinding> lookupFuzzyLocationInCountry(String location, String country, int results, String username) {
        // parameter check
        if (results < 0 || results > 100) {
            results = 10;
        }
        // build SPARQL query
        String url = URL_FUZZY + StringUtils.makeURLSafe(location) + VAR_COUNTRY + country + VAR_ROWS + results + VAR_SUFFIX + username;
        // send query
        String json = getResultsFromEndpoint(url);
        
        return parseJSON(json);
    }
    
    public static Map<String,GeonamesBinding> lookupFuzzyLocationInContinent(String location, String continent, int results, String username) {
        // parameter check
        if (results < 0 || results > 100) {
            results = 10;
        }
        // build SPARQL query
        String url = URL_FUZZY + StringUtils.makeURLSafe(location) + VAR_CONTINENT + continent + VAR_ROWS + results + VAR_SUFFIX + username;
        // send query
        String json = getResultsFromEndpoint(url);
        
        return parseJSON(json);
    }
}

File: src/clients/geonames-client.ts
/**
 * Geonames API client for geographic name lookups
 */

import { BaseSparqlClient } from './base-sparql-client.js';
import { GeonamesBinding, GeonamesConfig } from '../types/index.js';
import { makeURLSafe } from '../utils/string-utils.js';
import { LinkedDataError } from '../types/index.js';

interface GeonamesResult {
  geonameId: number;
  toponymName: string;
  countryCode: string;
  fcl: string;
  fcode: string;
}

interface GeonamesResponse {
  geonames: GeonamesResult[];
}

/**
 * Client for querying Geonames API
 */
export class GeonamesClient extends BaseSparqlClient {
  private static readonly URL_PRECISE = 'http://api.geonames.org/search?name_equals=';
  private static readonly URL_FUZZY = 'http://api.geonames.org/search?name=';
  private static readonly VAR_ROWS = '&maxRows=';
  private static readonly VAR_COUNTRY = '&country=';
  private static readonly VAR_FAVOR_UK = '&countryBias=GB';
  private static readonly VAR_CONTINENT = '&continentCode=';
  private static readonly VAR_SUFFIX =
    '&featureClass=A&featureClass=P&style=SHORT&type=json&username=';
  private static readonly URI_PREFIX = 'http://sws.geonames.org/';

  private username: string;

  constructor(config: GeonamesConfig) {
    super(config);

    if (!config.username) {
      throw new LinkedDataError('Geonames username is required');
    }

    this.username = config.username;
  }

  /**
   * Parses Geonames JSON response
   */
  private parseGeonamesJSON(json: string): Map<string, GeonamesBinding> {
    const results = new Map<string, GeonamesBinding>();

    try {
      const parsed: GeonamesResponse = JSON.parse(json);

      if (!parsed.geonames || !Array.isArray(parsed.geonames)) {
        return results;
      }

      for (const item of parsed.geonames) {
        const binding: GeonamesBinding = {
          uri: `${GeonamesClient.URI_PREFIX}${item.geonameId}`,
          label: item.toponymName,
          toponymName: item.toponymName,
          geonameId: String(item.geonameId),
          countryCode: item.countryCode,
          functionClass: item.fcl,
          functionCode: item.fcode,
        };

        results.set(binding.uri!, binding);
      }
    } catch (error) {
      console.error('Failed to parse Geonames JSON:', error);
    }

    return results;
  }

  /**
   * Looks up a precise location within a specific country
   */
  async lookupPreciseLocationInCountry(
    location: string,
    country: string,
    maxResults: number = 10
  ): Promise<Map<string, GeonamesBinding>> {
    const validatedResults = this.validateResultsCount(maxResults);

    const url =
      GeonamesClient.URL_PRECISE +
      makeURLSafe(location) +
      GeonamesClient.VAR_COUNTRY +
      country +
      GeonamesClient.VAR_ROWS +
      validatedResults +
      GeonamesClient.VAR_SUFFIX +
      this.username;

    const json = await this.getResultsFromEndpoint(url);
    return this.parseGeonamesJSON(json);
  }

  /**
   * Looks up a precise location within a specific continent
   */
  async lookupPreciseLocationInContinent(
    location: string,
    continent: string,
    maxResults: number = 10
  ): Promise<Map<string, GeonamesBinding>> {
    const validatedResults = this.validateResultsCount(maxResults);

    const url =
      GeonamesClient.URL_PRECISE +
      makeURLSafe(location) +
      GeonamesClient.VAR_CONTINENT +
      continent +
      GeonamesClient.VAR_ROWS +
      validatedResults +
      GeonamesClient.VAR_SUFFIX +
      this.username;

    const json = await this.getResultsFromEndpoint(url);
    return this.parseGeonamesJSON(json);
  }

  /**
   * Looks up a precise location anywhere in the world
   */
  async lookupPreciseLocationInWorld(
    location: string,
    maxResults: number = 10
  ): Promise<Map<string, GeonamesBinding>> {
    const validatedResults = this.validateResultsCount(maxResults);

    const url =
      GeonamesClient.URL_PRECISE +
      makeURLSafe(location) +
      GeonamesClient.VAR_ROWS +
      validatedResults +
      GeonamesClient.VAR_SUFFIX +
      this.username;

    const json = await this.getResultsFromEndpoint(url);
    return this.parseGeonamesJSON(json);
  }

  /**
   * Looks up a location with UK bias
   */
  async lookupLocationFavorUK(
    location: string,
    maxResults: number = 10
  ): Promise<Map<string, GeonamesBinding>> {
    const validatedResults = this.validateResultsCount(maxResults);

    const url =
      GeonamesClient.URL_PRECISE +
      makeURLSafe(location) +
      GeonamesClient.VAR_FAVOR_UK +
      GeonamesClient.VAR_ROWS +
      validatedResults +
      GeonamesClient.VAR_SUFFIX +
      this.username;

    const json = await this.getResultsFromEndpoint(url);
    return this.parseGeonamesJSON(json);
  }

  /**
   * Performs fuzzy location lookup within a specific country
   */
  async lookupFuzzyLocationInCountry(
    location: string,
    country: string,
    maxResults: number = 10
  ): Promise<Map<string, GeonamesBinding>> {
    const validatedResults = this.validateResultsCount(maxResults);

    const url =
      GeonamesClient.URL_FUZZY +
      makeURLSafe(location) +
      GeonamesClient.VAR_COUNTRY +
      country +
      GeonamesClient.VAR_ROWS +
      validatedResults +
      GeonamesClient.VAR_SUFFIX +
      this.username;

    const json = await this.getResultsFromEndpoint(url);
    return this.parseGeonamesJSON(json);
  }

  /**
   * Performs fuzzy location lookup within a specific continent
   */
  async lookupFuzzyLocationInContinent(
    location: string,
    continent: string,
    maxResults: number = 10
  ): Promise<Map<string, GeonamesBinding>> {
    const validatedResults = this.validateResultsCount(maxResults);

    const url =
      GeonamesClient.URL_FUZZY +
      makeURLSafe(location) +
      GeonamesClient.VAR_CONTINENT +
      continent +
      GeonamesClient.VAR_ROWS +
      validatedResults +
      GeonamesClient.VAR_SUFFIX +
      this.username;

    const json = await this.getResultsFromEndpoint(url);
    return this.parseGeonamesJSON(json);
  }
}
